whylog/constraints/__init__.py

```diff
@@ -1,52 +1,202 @@
+        2, None - at least 2 groups
+        2, 2 - exactly 2 groups
+        """
+        return cls.MIN_GROUPS_COUNT, cls.MAX_GROUPS_COUNT
 
-class DifferentValueIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
+    @abstractmethod
+    def verify(self, group_contents, param_dict):
+        """
+        Verifies constraint for given params in param_dict and groups contents.
+
+        :param param_dict: dict of additional params of constraint
+        :param groups: list of groups contents,
 
+        For LogReader and Teacher verification.
+        It must be optimized as well as possible (for LogReader).
+        """
 
-class ConstIntervals(AbstractConstraint):
-    def __init__(self, intervals):
         pass
 
 
-class ValueDeltaIntervals(AbstractConstraint):
-    def __init__(self, interval_lower, interval_greater, min_delta=None, max_delta=None):
+class TimeConstraint(AbstractConstraint):
+    """
+    Time delta between two dates must be greater or equal to 'min_delta'
+    and lower or equal to 'max_delta'
+    """
+
+    TYPE = ConstraintType.TIME_DELTA
+
+    MIN_GROUPS_COUNT = 2
+    MAX_GROUPS_COUNT = 2
+
+    MIN_DELTA = 'min_delta'
+    MAX_DELTA = 'max_delta'
+
+    PARAMS = sorted([MIN_DELTA, MAX_DELTA])
+
+    def __init__(self, groups=None, param_dict=None, params_checking=True):
         """
-        Sets Minimum and maximum difference between values of params (if values are numbers).
+        :param groups: First element of groups is a group with earlier date.
+
+        I.e:
+        TimeConstraint(
+            [(0, 1), (2, 1)]
+            {'min_delta': datetime.timedelta(seconds=1), 'max_delta': datetime.timedelta(seconds=10)},
+        )
         """
-        pass
+        super(TimeConstraint, self).__init__(groups, param_dict, params_checking)
+
+    def _check_optional_params(self, correct_param_names, actual_param_names):
+        if self.MIN_DELTA not in actual_param_names and self.MAX_DELTA not in actual_param_names:
+            raise ConstructorParamsError(self.TYPE, correct_param_names, actual_param_names)
+
+    def verify(self, group_contents, param_dict):
+        # TODO remove mock
+        return True
+
+
+class IdenticalConstraint(AbstractConstraint):
+    """
+    Contents of groups must be identical.
+    """
+
+    TYPE = ConstraintType.IDENTICAL
+
+    MIN_GROUPS_COUNT = 2
+    MAX_GROUPS_COUNT = None
+
+    PARAMS = []
+
+    def __init__(self, groups=None, param_dict=None, params_checking=True):
+        """
+        I.e:
+        IdenticalConstraint(
+            [(1, 2), (2, 4)]
+        )
+        """
+        super(IdenticalConstraint, self).__init__(groups, param_dict, params_checking)
+
+    def verify(self, group_contents, param_dict=None):
+        """
+        I.e:
+        - verify({}, ['comp1', 'comp1', 'comp1']) returns True
+        - verify({}, ['comp1', 'hello', 'comp1']) returns False
+        """
+
+        if len(group_contents) <= 1:
```

Comment:
remove that `if` block. This condition is already checked by `AbstractConstraint._check_constructor_groups()` once per life of constraint, not once per `verify()` call. This approach is like a billion times more effective.